server/app/services/ProgramBlockValidationFactory.java

```diff
@@ -0,0 +1,24 @@
+package services;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import javax.inject.Inject;
+import services.ProgramBlockValidation.AddQuestionResult;
+import services.program.BlockDefinition;
+import services.program.ProgramDefinition;
+import services.question.types.QuestionDefinition;
+
+public final class ProgramBlockValidationFactory {
+
+  private final ProgramBlockValidation programBlockValidation;
+
+  @Inject
+  public ProgramBlockValidationFactory(ProgramBlockValidation programBlockValidation) {
+    this.programBlockValidation = checkNotNull(programBlockValidation);
+  }
+
+  public AddQuestionResult canAddQuestion(
+      ProgramDefinition program, BlockDefinition block, QuestionDefinition question) {
+    return programBlockValidation.canAddQuestion(program, block, question);
+  }
```

Comment:
**suggestion (blocking):** This is still the same behavior as before, just with another class in between. The goal is to avoid calling database query methods in a loop and instead call them only once. In this case, the database query method we want to avoid calling repeatedly is inside `versionRepository.getDraftVersion()`.

Factory classes should have a single method that creates a new instance of the class they're a factory for. In this case, `ProgramBlockValidationFactory` should have a `create` method that returns a new instance of `ProgramBlockValidation`:
```suggestion
  @Inject
  public ProgramBlockValidationFactory(VersionRepository versionRepository) {
    this.versionRepository = checkNotNull(versionRepository);
  }

  public ProgramBlockValidation create() {
    Version currentDraft = versionRepository.getDraftVersion();
    return new ProgramBlockValidation(currentDraft);
  }
```

Then, in `ProgramQuestionBank.java`:
```java
  private Stream<QuestionDefinition> filterQuestions() {
    ProgramBlockValidation programBlockValidator = programBlockValidationFactory.create();
    return params.questions().stream()
        .filter(
            q ->
                programBlockValidator.canAddQuestion(params.program(), params.blockDefinition(), q)
                    == ProgramBlockValidation.AddQuestionResult.ELIGIBLE);
                programBlockValidator.canAddQuestion(
                        params.program(), params.blockDefinition(), q)
                    == AddQuestionResult.ELIGIBLE);
  }
  ```

Reply:
This make sense. Thank you for clarifying